README.md
# pathdiagram

The R package **pathdiagram** implements simple functions to draw basic PLS path diagrams just for visualization purposes.

## Installation

To install the prototype version on [github](https://github.com/gastonstat/pathdiagram) using the package **devtools**, simply run in R:
```
# install.packages("devtools") 
library(devtools)
install_github('pathdiagram',  username='gastonstat')
```
More info at http://www.gastonsanchez.com/pathdiagram


## Motivation

The one and only reason to create **pathdiagram** was the need to plot a path diagram in R. Since I was writing the tutorials and demos for my other packages **plspm** and **plsdepot**, I realized that I badly needed path diagrams accompanying my examples. I first tried using the **grid** package. It was fine but it didn't allow me to have complete control over the looking of the objects (mainly with color of lines, borders, text, etc). Then I tried to use the package **diagram** but I got the same restrictions. Finally, after some short experiments, I decided to create **pathdiagram**.  


## devtools reminder

Documentation cycle steps using devtools
More info at https://github.com/hadley/devtools
```
# load packages
library(devtools)
library(digest)
library(roxygen2)
# generate documentation files
document(pkg="pathdiagram")
# check documentation
check_doc(pkg="pathdiagram")
``` 


NEWS
Version 0.1.8 (2012-11-08)
-----------------------------------------

* remove parameter `mar` from function `wall` (it wasn't that useful)

* add parameter `xpd` to function `wall`


R/draw.R
#'@title Draw manifest and latent variables
#'
#'@description
#'Use this function to draw either manifest or latent variables on a plot.
#'
#'@param obj An object of either class \code{"manifest"} or \code{"latent"}
#'@author Gaston Sanchez
#'@seealso \code{\link{manifest}}, \code{\link{latent}}
#'@export
#'@examples
#'
#'  \dontrun{
#'  # manifest variables
#'  ingredients = list(
#'    eggs = manifest("eggs", x=0.3, y=0.7, width=0.10, height=0.08),
#'    milk = manifest("milk", x=0.3, y=0.6, width=0.10, height=0.08),
#'    flour = manifest("flour", x=0.3, y=0.5, width=0.10, height=0.08),
#'    sugar = manifest("sugar", x=0.3, y=0.4, width=0.10, height=0.08),
#'    butter = manifest("butter", x=0.3, y=0.3, width=0.10, height=0.08)
#'  )
#'
#'  # latent variables
#'  pancakes = latent("PANCAKES", x=0.6, y=0.6, rx=0.09, ry=0.07)
#'  waffles = latent("WAFFLES", x=0.6, y=0.4, rx=0.09, ry=0.07)
#'
#'  # open wall
#'  wall()
#'  
#'  title("Toy Path Diagram", col.main="gray20")
#'  # draw manifest variables
#'  for (i in 1:length(ingredients)) {
#'     draw(ingredients[[i]])
#'  }
#'  
#'  # draw latent variables
#'  draw(pancakes)
#'  draw(waffles)
#'  # draw arrows
#'  for (i in 1:length(ingredients)) {
#'     arrow(from=ingredients[[i]], to=pancakes, start="east", end="west")
#'     arrow(from=ingredients[[i]], to=waffles, start="east", end="west")
#'  }
#'  }
#'  
draw <- 
function(obj)
{
  if (class(obj) == "manifest")
  {
    rect(obj$xl, obj$yb, obj$xr, obj$yt, col=obj$fill, border=obj$border,
         lwd=obj$lwd)
    text(obj$north[1], obj$east[2], labels=obj$label, col=obj$col,
         cex=obj$cex, vfont=obj$vfont, font=obj$font, family=obj$family)    
  }
  if (class(obj) == "latent")
  {
    plotellipse(rx=obj$rx, ry=obj$ry, mid=c(obj$x, obj$y), 
                lcol=obj$lcol, col=obj$fill, lwd=obj$lwd)
    text(obj$x, obj$y, labels=obj$label, col=obj$col,
         cex=obj$cex, vfont=obj$vfont, font=obj$font, family=obj$family)
  }
}


R/wall.R
#'@title Open a new frame for a path diagram
#'
#'@description
#'Use this function to open a white canvas to start drawing a path diagram. 
#'By default, \code{wall} opens a new plot window from 0 to 1 in both axes.
#'
#'@details
#'\code{wall} calls \code{plot.new()} and \code{plot.window()} to open a new plot frame.
#'
#'@param xlim Numeric vector of length 2 giving the x coordinate range.
#'Default \code{xlim = c(0, 1)}.
#'@param ylim Numeric vector of length 2 giving the y coordinate range.
#'Default \code{ylim = c(0, 1)}.
#'@param xpd Logical value to indicate if all plotting is clipped to the figure region. 
#'The default is c(1, 1, 1, 1).
#'@param ... other graphical arguments passed on to \code{\link{plot.window}}.
#'@author Gaston Sanchez
#'@seealso \code{\link{manifest}}, \code{\link{latent}}, \code{\link{draw}}
#'@export
#'@examples
#'
#'  \dontrun{
#'  # latent variables
#'  attack = latent("ATTACK", x=0.35, y=0.7, rx=0.08, ry=0.06)
#'  defense = latent("DEFENSE", x=0.35, y=0.3, rx=0.08, ry=0.06)
#'  success = latent("SUCCESS", x=0.65, y=0.5, rx=0.08, ry=0.06)
#'
#'  # open diagram
#'  wall()
#'  
#'  # draw latent variables
#'  draw(attack)
#'  draw(defense)
#'  draw(success)
#'  
#'  # add arrows
#'  arrow(from=attack, to=success, start="east", end="west")
#'  arrow(from=defense, to=success, start="east", end="west")
#'  }
#'
wall <-
function(xlim = c(0, 1), ylim = c(0, 1), xpd = TRUE, ...)
{
  plot.new()
  plot.window(xlim = xlim, ylim = ylim, xpd = xpd, ...)
}


R/arrow.R
#'@title Plot arrow between variables
#'
#'@description
#'Use this function to draw connecting arrows between manifest and latent variables.
#'
#'@param from An object of either class \code{"manifest"} or \code{"latent"}. This
#'object is the origin of the arrow. 
#'@param to An object of either class \code{"manifest"} or \code{"latent"}. This
#'object is the destination of the arrow.
#'@param start Character string to specify the starting direction of the arrow. Options
#'are \code{"north"}, \code{"south"}, \code{"east"}, \code{"west"}.
#'@param end Character string to specify the ending direction of the arrow. Options
#'are \code{"north"}, \code{"south"}, \code{"east"}, \code{"west"}.
#'@param length length of the edges of the arrow head (in inches).
#'@param angle angle from the shaft of the arrow to the edge of the arrow head.
#'@param code integer code, determining kind of arrows to be drawn.
#'@param col color of the arrow.
#'@param lwd width of the arrow.
#'@param ... other arguments passed on to \code{arrows}.
#'@author Gaston Sanchez
#'@seealso \code{\link{manifest}}, \code{\link{latent}}, \code{\link{draw}}
#'@export
#'@examples
#'
#'  \dontrun{
#'  # latent variables
#'  attack = latent("ATTACK", x=0.35, y=0.7, rx=0.08, ry=0.06)
#'  defense = latent("DEFENSE", x=0.35, y=0.3, rx=0.08, ry=0.06)
#'  success = latent("SUCCESS", x=0.65, y=0.5, rx=0.08, ry=0.06)
#'
#'  # open wall
#'  wall()
#'  # draw latent variables
#'  draw(attack)
#'  draw(defense)
#'  draw(success)
#'  # add arrows
#'  arrow(from=attack, to=success, start="east", end="west")
#'  arrow(from=defense, to=success, start="east", end="west")
#'  }
#'
arrow <- 
function(from, to, start = "east", end = "west", length = 0.1,
         angle = 10, code = 2, col = "#d2def1", lwd = 3, ...)
{
  ini = grep(start, c("north", "south", "east", "west")) + 5
  fin = grep(end, c("north", "south", "east", "west")) + 5
  x0 = from[[ini]][1]
  y0 = from[[ini]][2]  
  x1 = to[[fin]][1]
  y1 = to[[fin]][2]
  arrows(x0=x0, y0=y0, x1=x1, y1=y1, length=length, angle=angle, 
         code=code, col=col, lwd=lwd, ...)
}


R/latent.R
#'@title Set specifications of a latent variable
#'
#'@description
#'Use this function to specify the graphic characteristics of a latent variable.
#'The specifications will be used by the function \code{draw} to plot latent
#'variables (in a path diagram).
#'
#'@details
#'Latent variables are drawn as ellipses using the function \code{\link{plotellipse}}
#'
#'@param label A character string with the label to be displayed.
#'@param x x-axis coordinate for center of ellipse.
#'@param y y-axis coordinate for center of ellipse.
#'@param rx long radius of ellipse.
#'@param ry short radius of ellipse.
#'@param border color of the border.
#'@param lwd width of border line.
#'@param fill color to fill the ellipse
#'@param col color of the label.
#'@param cex numeric character expansion of the label. 
#'@param vfont font family of the label.
#'@param font An integer specifying which font to use for the label.
#'See \code{\link{par}}
#'@param family The name of a font family for drawing text. 
#'Standard values are \code{"serif"}, \code{"sans"} and \code{"mono"}.
#'@return An object of class \code{"latent"}, which is a list with the
#'specified parameters to draw latent variables.
#'@author Gaston Sanchez
#'@seealso \code{\link{manifest}}, \code{\link{draw}}
#'@export
#'@examples
#'
#'  \dontrun{
#'  # latent variables
#'  attack = latent("ATTACK", x=0.35, y=0.7, rx=0.08, ry=0.06)
#'  defense = latent("DEFENSE", x=0.35, y=0.3, rx=0.08, ry=0.06)
#'  success = latent("SUCCESS", x=0.65, y=0.5, rx=0.08, ry=0.06)
#'
#'  # opwn wall
#'  wall()
#'  title("Drawing three latent variables", col.main="gray20")
#'  
#'  # draw variables
#'  draw(attack)
#'  draw(defense)
#'  draw(success)
#'  }
#'
latent <- 
  function(label = "latent", x = 0.5, y = 0.5, rx = 0.05, ry = 0.05,
           border = "white", lwd = 2, fill = "#5f8bd7", col = "white",
           cex = 1, vfont = NULL, font = 2, family = "sans")
{
  # list of specs
  lat = list(label = label, 
             x = x, 
             y = y, 
             rx = rx, 
             ry = ry,
             north = c(x, y+ry),
             south = c(x, y-ry),
             east = c(x+rx, y), 
             west = c(x-rx, y),
             lcol = border, 
             fill = fill,
             lwd = lwd,
             col = col,
             cex = cex,
             font = font,
             vfont = vfont,
             family = family)
  class(lat) = "latent"
  lat
}


R/manifest.R
#'@title Set specifications of a manifest variable
#'
#'@description
#'Use this function to specify the graphic characteristics of a manifest variable.
#'The specifications will be used by the function \code{draw} to plot manifest
#'variables (in a path diagram).
#'
#'@details
#'Manifest variables are drawn as rectangles.
#'
#'@param label A character string with the label to be displayed.
#'@param x x-axis coordinate for center of rectanlge.
#'@param y y-axis coordinate for center of rectangle.
#'@param width width of the rectangle.
#'@param height height of the rectangle.
#'@param border color of the border.
#'@param fill color to fill the rectangle.
#'@param lwd width of the border.
#'@param col color of the label.
#'@param cex numeric character expansion of the label. 
#'@param vfont font family of the label.
#'@param font An integer specifying which font to use for the label.
#'See \code{\link{par}}
#'@param family The name of a font family for drawing text. 
#'Standard values are \code{"serif"}, \code{"sans"} and \code{"mono"}.
#'@return An object of class \code{"manifest"}, which is a list with the
#'specified parameters to draw manifest variables.
#'@author Gaston Sanchez
#'@seealso \code{\link{latent}}, \code{\link{draw}}
#'@export
#'@examples
#'
#'  \dontrun{
#'  # manifest variables
#'  eggs = manifest("eggs", x=0.3, y=0.7, width=NULL, height=0.08)
#'  milk = manifest("milk", x=0.4, y=0.6, width=NULL, height=0.08)
#'  flour = manifest("flour", x=0.5, y=0.5, width=NULL, height=0.08)
#'  sugar = manifest("sugar", x=0.6, y=0.4, width=NULL, height=0.08)
#'  butter = manifest("butter", x=0.7, y=0.3, width=NULL, height=0.08)
#'
#'  # open wall
#'  wall()
#'  title("Five manifest variables", col.main="gray20")
#'  
#'  # draw manifest variables
#'  draw(eggs)
#'  draw(milk)
#'  draw(flour)
#'  draw(sugar)
#'  draw(butter)  
#'  }
#'
manifest <- 
  function(label = "manifest", x = 0.5, y = 0.5, width = NULL, height = 0.1,
           border = "white", fill = "#9dbafa", lwd = 1, col = "gray20", cex = 1,
           vfont = NULL, font = 1, family = "sans")
{
  if (is.null(width))
    width = nchar(label) / 50
  xl = x - width/2
  yb = y - height/2
  xr = x + width/2
  yt = y + height/2
  # list of specs
  man = list(label = label, 
             xl = xl, 
             yb = yb, 
             xr = xr, 
             yt = yt,
             north = c(x, yt),
             south = c(x, yb),
             east = c(xr, y), 
             west = c(xl, y),
             border = border,
             fill = fill,
             lwd = lwd,
             col = col,
             cex = cex,
             vfont = vfont,
             font = font,
             family = family)
  class(man) = "manifest"
  man
}
